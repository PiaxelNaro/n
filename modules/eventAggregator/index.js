// @flow
const makeEventAggregator = (target: HTMLElement, event: string, getValue: Function) => {
    let currentValue = NaN;
    let lastSentValue = NaN;
    let listenerAdded = false;
    let handlers = [];

    const eventHandler = (...args) => {
        currentValue = getValue(...args);
    };

    const loopHandlers = () => {
        if (!handlers.length) {
            return;
        }

        if (lastSentValue !== currentValue) {
            lastSentValue = currentValue;
            handlers.forEach(handler => handler(currentValue));
        }

        global.requestAnimationFrame(loopHandlers);
    };

    const startWatching = () => {
        if (listenerAdded) {
            return true;
        }

        listenerAdded = true;
        eventHandler();
        target.addEventListener(event, eventHandler);
        global.requestAnimationFrame(loopHandlers);
        return false;
    };

    const stopWatching = () => {
        target.removeEventListener(event, eventHandler);
        lastSentValue = NaN;
        currentValue = NaN;
        listenerAdded = false;
        handlers = [];
    };

    return (handler: Function) => {
        handlers.push(handler);
        if (startWatching()) {
            handler(currentValue);
        }

        return () => {
            // Remove handlers from
            handlers = handlers.filter(item => item !== handler);

            // Stop watcher if there's no handlers left
            if (!handlers.length) {
                stopWatching();
            }
        };
    };
};

export default makeEventAggregator;

export const scrollAggregator = makeEventAggregator(
    global.document,
    'scroll',
    () => global.document.body.scrollTop || global.document.documentElement.scrollTop
);

export const resizeAggregator = makeEventAggregator(
    global.window,
    'resize',
    () => global.document.body.offsetWidth
);
