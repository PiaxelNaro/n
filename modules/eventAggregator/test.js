// @flow
import makeEventAggregator, { scrollAggregator, resizeAggregator } from '.';

describe('makeEventAggregator()', () => {
    const getValue = jest.fn(() => 'a');
    global.requestAnimationFrame = jest.fn();
    const target: any = {
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
    };
    const aggregator = makeEventAggregator(target, 'my-event', getValue);

    // This module is lazy, it'll only add listeners when required.
    expect(getValue).not.toHaveBeenCalled();
    expect(target.addEventListener).not.toHaveBeenCalled();

    const cb1 = jest.fn();
    const unsubscribe1 = aggregator(cb1);
    expect(target.addEventListener).toHaveBeenCalled();

    const cb2 = jest.fn();
    const unsubscribe2 = aggregator(cb2);
    expect(target.addEventListener).toHaveBeenCalledTimes(1);

    const eventHandler = target.addEventListener.mock.calls[0][1];
    const rAF = global.requestAnimationFrame.mock.calls[0][0];

    // Multiple events will be aggregated
    eventHandler('arg');
    expect(getValue).toHaveBeenCalledWith('arg');
    eventHandler();

    expect(cb1).not.toHaveBeenCalled();

    // On the next animation frame, we'll call the handlers and schedule the next check for the
    // next frame.
    rAF();
    expect(cb1).toHaveBeenCalledWith('a');
    expect(cb2).toHaveBeenCalledWith('a');
    expect(global.requestAnimationFrame).toHaveBeenCalledTimes(2);

    // Handlers are not called when value doesn't change
    eventHandler();
    rAF();
    expect(cb1).toHaveBeenCalledTimes(1);

    // Clean itself on unsubscribe
    unsubscribe1();
    expect(target.removeEventListener).not.toHaveBeenCalled();
    unsubscribe2();
    expect(target.removeEventListener).toHaveBeenCalled();
});

describe('scrollAggregator', () => {
    it('aggregate document scroll events', () => {
        const spyFirstCallBack = () => 1;
        const spySecondCallBack = () => 2;
        global.requestAnimationFrame = jest.fn();
        global.document.addEventListener = jest.fn();
        global.document.removeEventListener = jest.fn();
        const unmountFirst = scrollAggregator(spyFirstCallBack);
        const scrollHandler = global.document.addEventListener.mock.calls[0][1];
        const loopHandlers = global.requestAnimationFrame.mock.calls[0][0];
        scrollHandler();
        expect(global.document.addEventListener).toHaveBeenCalledTimes(1);

        global.document.body.scrollTop = 1;
        scrollHandler();
        loopHandlers();
        expect(global.requestAnimationFrame).toHaveBeenCalledTimes(2);

        const unmountSecond = scrollAggregator(spySecondCallBack);
        unmountFirst();
        loopHandlers();
        expect(global.requestAnimationFrame).toHaveBeenCalledTimes(3);
        expect(global.document.removeEventListener).toHaveBeenCalledTimes(0);

        unmountSecond();
        loopHandlers();
        expect(global.document.removeEventListener).toHaveBeenCalledTimes(1);
    });
});

describe('resizeAggregator', () => {
    it('aggregate window resize events', () => {
        const spy = jest.fn();
        global.requestAnimationFrame = jest.fn();
        global.window.addEventListener = jest.fn();
        global.window.removeEventListener = jest.fn();

        const unsubscribe = resizeAggregator(spy);

        const triggerResize = global.window.addEventListener.mock.calls[0][1];
        const triggerRaF = global.requestAnimationFrame.mock.calls[0][0];

        expect(global.window.addEventListener).toHaveBeenCalledTimes(1);

        triggerResize();
        triggerRaF();
        expect(spy).toHaveBeenCalledTimes(1);
        unsubscribe();
    });
});
