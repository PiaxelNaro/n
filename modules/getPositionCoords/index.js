// @flow
export type PositionningCoords = {
    style: {
        top: number,
        left: number,
        width: number,
    },
    alignTo: string,
};

/**
 * Returns the coordinates an element must use in order to be positionned under a target.
 */
export default function getPositionCoords(
    target: HTMLElement,
    win: window = global
): PositionningCoords {
    const box = target.getBoundingClientRect();

    const body = win.document.body;
    const docEl = win.document.documentElement;

    const scrollTop = win.pageYOffset || docEl.scrollTop || body.scrollTop;
    const scrollLeft = win.pageXOffset || docEl.scrollLeft || body.scrollLeft;

    const clientTop = docEl.clientTop || body.clientTop || 0;
    const clientLeft = docEl.clientLeft || body.clientLeft || 0;
    const clientWidth = docEl.clientWidth || body.clientWidth || 0;

    const top = box.top + box.height + scrollTop - clientTop;
    const left = box.left + scrollLeft - clientLeft;

    // Align right
    const tiers = clientWidth / 3;
    return {
        style: {
            top: Math.round(top),
            left: Math.round(left),
            width: box.width,
        },
        alignTo: left / tiers > 2 ? 'right' : 'left',
    };
}
