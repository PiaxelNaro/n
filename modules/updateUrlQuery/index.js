// @flow
import qs from 'querystring';
import nextRouter from 'next/router';

const Router: any = nextRouter;

export default (
    params: Object,
    options: { replace?: boolean, shallow?: boolean } = { replace: false, shallow: false }
) => {
    const url = `${Router.pathname}?${qs.stringify(params)}`;
    if (options.replace) {
        Router.replace(url, url, { shallow: options.shallow });
    } else {
        Router.push(url, url, { shallow: options.shallow });
    }
};
