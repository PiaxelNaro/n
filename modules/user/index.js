// @flow
export type UserType = {
    profile: {
        surveyFirstName?: string,
        surveyName: string,
        email: string,
        fileRepository?: string,
    },
    isANA: boolean,
    isINT: boolean,
    isApiPartner?: boolean,
    intAuth: {
        store?: {
            pp: boolean,
            pya: boolean,
            pyp: boolean,
        },
        usage?: {
            pp: boolean,
            pya: boolean,
        },
        marketing?: {
            advertising: boolean,
            aso: boolean,
        },
    },
};
