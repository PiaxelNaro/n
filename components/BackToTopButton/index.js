// @flow
import React from 'react';
import styled from 'styled-components';
import ChevronUpIcon from 'react-icons/lib/fa/chevron-up';
import { scrollAggregator } from '../../modules/eventAggregator';

const Wrapper = styled.div`
    width: 32px;
    height: 32px;
    line-height: 26px;
    position: fixed;
    bottom: 48px;
    right: 30px;
    cursor: pointer;
    border-radius: 2px;
    text-align: center;
    box-shadow: 0 0 2px #61b9e8;
    background: rgba(10, 10, 10, 0.3);
    color: rgba(255, 255, 255, 0.8);
    &:hover,
    &:focus {
        background: rgba(10, 10, 10, 0.5);
        color: #fff;
    }
`;

const backToTop = () => {
    global.document.body.scrollTop = 0; // For Chrome, Safari and Opera
    global.document.documentElement.scrollTop = 0; // For IE and Firefox
};

class BackToTopButton extends React.Component {
    state: {
        display: boolean,
    };

    scrollWatcher: Function;
    removeScrollWatcher: Function;

    constructor() {
        super();

        this.state = {
            display: false,
        };

        this.scrollWatcher = this.scrollWatcher.bind(this);
    }

    componentDidMount() {
        this.removeScrollWatcher = scrollAggregator(this.scrollWatcher);
    }

    componentWillUnmount() {
        this.removeScrollWatcher();
    }

    scrollWatcher(scrollTop: number) {
        this.setState({ display: scrollTop > 0 });
    }

    render() {
        if (!this.state.display) {
            return null;
        }

        return (
            <Wrapper onClick={backToTop}>
                <ChevronUpIcon size={20} />
            </Wrapper>
        );
    }
}

export default BackToTopButton;
