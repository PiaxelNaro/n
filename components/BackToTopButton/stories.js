// @flow
import React from 'react';
import { storiesOf } from '@kadira/storybook';
import BackToTopButton from '.';

storiesOf('BackToTopButton', module).add('render back to top button', () => (
    <div style={{ height: '300%' }}>
        <h1> Do You Know?</h1>
        <h4> Bo probably is one of the happiest person in </h4>
        <h1> Do You Know?</h1>
        <h4> Bo probably is one of the happiest person in </h4>
        <h1> Do You Know?</h1>
        <h4> Bo probably is one of the happiest person in </h4>
        <h1> Do You Know?</h1>
        <h4> Bo probably is one of the happiest person in </h4>
        <h1> Do You Know?</h1>
        <h4> Bo probably is one of the happiest person in </h4>
        <h1> Do You Know?</h1>
        <h4> Bo probably is one of the happiest person in </h4>
        <h1> Do You Know?</h1>
        <h4> Bo probably is one of the happiest person in </h4>
        <h1> Do You Know?</h1>
        <h4> Bo probably is one of the happiest person in </h4>
        <h1> Do You Know?</h1>
        <h4> Bo probably is one of the happiest person in </h4>
        <h1> Do You Know?</h1>
        <h4> Bo probably is one of the happiest person in </h4>
        <h1> Do You Know?</h1>
        <h4> Bo probably is one of the happiest person in </h4>
        <h1> Do You Know?</h1>
        <h4> Bo probably is one of the happiest person in </h4>
        <h1> Do You Know?</h1>
        <h4> Bo probably is one of the happiest person in </h4>
        <h1> Do You Know?</h1>
        <h4> Bo probably is one of the happiest person in </h4>
        <h1> Do You Know?</h1>
        <h4> Bo probably is one of the happiest person in </h4>
        <BackToTopButton />
    </div>
));
