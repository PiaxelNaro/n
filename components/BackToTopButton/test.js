// @flow
import React from 'react';
import { shallow, mount } from 'enzyme';
import BackToTopButton from '.';
import * as eventAggregator from '../../modules/eventAggregator';

const scrollAggregator: any = eventAggregator.scrollAggregator;

jest.mock('../../modules/eventAggregator', () => {
    const unsubscribe = jest.fn();
    return {
        scrollAggregator: jest.fn(() => unsubscribe),
    };
});

describe('<BackToTopButton>', () => {
    it('scroll to top button should display if user scroll down', () => {
        const button = shallow(<BackToTopButton />);
        expect(button).toMatchSnapshot();

        // Scroll down
        const inst: any = button.instance();
        inst.scrollWatcher(1);
        button.update();
        expect(button).toMatchSnapshot();

        // Scroll up
        inst.scrollWatcher(0);
        button.update();
        expect(button).toMatchSnapshot();
    });

    it('scroll to top on click', () => {
        global.document.body.scrollTop = 10;
        global.document.documentElement.scrollTop = 10;
        const button = mount(<BackToTopButton />);
        expect(scrollAggregator).toHaveBeenCalledTimes(1);

        // Trigger a scroll event
        scrollAggregator.mock.calls[0][0](10);

        button.simulate('click');
        expect(global.document.body.scrollTop).toEqual(0);
        expect(global.document.documentElement.scrollTop).toEqual(0);

        // Unmounting will clean the listeners.
        button.unmount();
        expect(scrollAggregator(jest.fn())).toHaveBeenCalledTimes(1);
    });
});
