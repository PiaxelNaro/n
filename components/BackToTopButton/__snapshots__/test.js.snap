// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BackToTopButton> scroll to top button should display if user scroll down 1`] = `null`;

exports[`<BackToTopButton> scroll to top button should display if user scroll down 2`] = `
<BackToTopButton__Wrapper
  onClick={[Function]}
>
  <FaChevronUp
    size={20}
  />
</BackToTopButton__Wrapper>
`;

exports[`<BackToTopButton> scroll to top button should display if user scroll down 3`] = `null`;
